File: testutils/src/main/java/androidx/testutils/LifecycleOwnerUtils.java
/*
 * Copyright 2019 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.testutils;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;

import android.app.Activity;
import android.app.Instrumentation;

import androidx.annotation.NonNull;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleEventObserver;
import androidx.lifecycle.LifecycleOwner;
import androidx.test.platform.app.InstrumentationRegistry;
import androidx.test.rule.ActivityTestRule;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

/**
 * Utility methods for testing LifecycleOwners
 */
public class LifecycleOwnerUtils {
    private static final long TIMEOUT_MS = 5000;

    private static final Runnable DO_NOTHING = new Runnable() {
        @Override
        public void run() {
        }
    };

    /**
     * Waits until the the Activity current held the ActivityTestRule has the specified
     * {@link androidx.lifecycle.Lifecycle.State}. If the owner has not hit that state within a
     * suitable time period, it asserts that the current state equals the given state.
     */
    public static <T extends Activity & LifecycleOwner> void waitUntilState(
            final ActivityTestRule<T> activityRule,
            final Lifecycle.State state) throws Throwable {
        waitUntilState(activityRule.getActivity(), activityRule, state);
    }

    /**
     * Waits until the given {@link LifecycleOwner} has the specified
     * {@link androidx.lifecycle.Lifecycle.State}. If the owner has not hit that state within a
     * suitable time period, it asserts that the current state equals the given state.
     */
    public static void waitUntilState(final LifecycleOwner owner,
            final ActivityTestRule<?> activityRule,
            final Lifecycle.State state) throws Throwable {
        final CountDownLatch latch = new CountDownLatch(1);
        activityRule.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                final Lifecycle.State currentState = owner.getLifecycle().getCurrentState();
                if (currentState == state) {
                    latch.countDown();
                    return;
                }
                owner.getLifecycle().addObserver(new LifecycleEventObserver() {
                    @Override
                    public void onStateChanged(@NonNull LifecycleOwner provider,
                            @NonNull Lifecycle.Event event) {
                        if (provider.getLifecycle().getCurrentState() == state) {
                            latch.countDown();
                            provider.getLifecycle().removeObserver(this);
                        }
                    }
                });
            }
        });
        final boolean latchResult = latch.await(30, TimeUnit.SECONDS);

        assertThat("Expected " + state + " never happened to " + owner
                        + ". Current state:" + owner.getLifecycle().getCurrentState(),
                latchResult,
                is(true));

        // wait for another loop to ensure all observers are called
        activityRule.runOnUiThread(DO_NOTHING);
    }

    /**
     * Waits until the given the current {@link Activity} has been recreated, and
     * the new instance is resumed.
     */
    @SuppressWarnings("unchecked")
    public static <T extends Activity & LifecycleOwner> T waitForRecreation(
            final ActivityTestRule<T> activityRule
    ) throws Throwable {
        return waitForRecreation(activityRule.getActivity(), activityRule);
    }

    /**
     * Waits until the given {@link Activity} and {@link LifecycleOwner} has been recreated, and
     * the new instance is resumed.
     */
    @SuppressWarnings("unchecked")
    public static <T extends Activity & LifecycleOwner> T waitForRecreation(
            final T activity,
            final ActivityTestRule<?> activityRule
    ) throws Throwable {
        Instrumentation.ActivityMonitor monitor = new Instrumentation.ActivityMonitor(
                activity.getClass().getCanonicalName(), null, false);
        Instrumentation instrumentation = InstrumentationRegistry.getInstrumentation();
        instrumentation.addMonitor(monitor);

        T result;

        // this guarantee that we will reinstall monitor between notifications about onDestroy
        // and onCreate
        //noinspection SynchronizationOnLocalVariableOrMethodParameter
        synchronized (monitor) {
            do {
                // The documentation says "Block until an Activity is created
                // that matches this monitor." This statement is true, but there are some other
                // true statements like: "Block until an Activity is destroyed" or
                // "Block until an Activity is resumed"...
                // this call will release synchronization monitor's monitor
                result = (T) monitor.waitForActivityWithTimeout(TIMEOUT_MS);
                if (result == null) {
                    throw new RuntimeException("Timeout. Activity was not recreated.");
                }
            } while (result == activity);
        }

        // Finally wait for the recreated Activity to be resumed
        waitUntilState(result, activityRule, Lifecycle.State.RESUMED);

        return result;
    }

    private LifecycleOwnerUtils() {
    }
}

File: testutils/testutils-runtime/src/main/java/androidx/testutils/LifecycleOwnerUtils.kt
/*
 * Copyright 2019 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package androidx.testutils

import android.app.Activity
import android.app.Instrumentation.ActivityMonitor
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver
import androidx.lifecycle.LifecycleOwner
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.rule.ActivityTestRule
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import org.hamcrest.CoreMatchers
import org.hamcrest.MatcherAssert

/** Utility methods for testing LifecycleOwners */
object LifecycleOwnerUtils {

    private const val TIMEOUT_MS: Long = 5000

    private val DO_NOTHING = Runnable {}

    /**
     * Waits until the given [LifecycleOwner] has the specified
     * [androidx.lifecycle.Lifecycle.State]. If the owner has not hit that state within a suitable
     * time period, it asserts that the current state equals the given state.
     */
    @JvmStatic
    @Throws(Throwable::class)
    fun waitUntilState(owner: LifecycleOwner, state: Lifecycle.State) {
        if (owner.lifecycle.currentState == state) {
            return
        }

        val latch = CountDownLatch(1)
        InstrumentationRegistry.getInstrumentation().runOnMainSync(
            Runnable {
                val currentState = owner.lifecycle.currentState
                if (currentState == state) {
                    latch.countDown()
                    return@Runnable
                }

            val observer =
                object : LifecycleEventObserver {
                    override fun onStateChanged(provider: LifecycleOwner, event: Lifecycle.Event) {
                        if (provider.lifecycle.currentState == state) {
                            provider.lifecycle.removeObserver(this)
                            latch.countDown()
                        }
                    }
                }

                owner.lifecycle.addObserver(observer)
            })
        val latchResult = latch.await(15, TimeUnit.SECONDS)

        MatcherAssert.assertThat(
            "Expected state $state never happened to $owner. Current state: ${owner.lifecycle.currentState}",
            latchResult,
            CoreMatchers.`is`(true)
        )

        // wait for another loop to ensure all observers are called
        InstrumentationRegistry.getInstrumentation().runOnMainSync(DO_NOTHING)
    }

    /**
     * Waits until the given the current [Activity] has been recreated, and the new instance is
     * resumed.
     */
    @Throws(Throwable::class)
    fun <T> waitForRecreation(
        @Suppress("deprecation") activityRule: ActivityTestRule<T>
    ): T where T : Activity, T : LifecycleOwner {
        return waitForRecreation(activityRule.activity)
    }

    /**
     * Waits until the given the given [Activity] has been recreated, and the new instance is
     * resumed.
     */
    @Throws(Throwable::class)
    fun <T> waitForRecreation(activity: T): T where T : Activity, T : LifecycleOwner {
        return waitForRecreation(activity, null)
    }

    /**
     * Waits until the given [Activity] and [LifecycleOwner] has been recreated, and the new
     * instance is resumed.
     */
    @Throws(Throwable::class)
    fun <T> waitForRecreation(activity: T, actionOnUiThread: Runnable?): T where
    T : Activity,
    T : LifecycleOwner {
        val monitor = ActivityMonitor(activity::class.qualifiedName, null, false)
        val instrumentation = InstrumentationRegistry.getInstrumentation()
        instrumentation.addMonitor(monitor)

        if (actionOnUiThread != null) {
            instrumentation.runOnMainSync(actionOnUiThread)
        }

        // Wait for the old activity to be destroyed. This helps avoid flakiness on test devices
        // (ex. API 26) where the system takes a long time to go from STOPPED to DESTROYED.
        waitUntilState(activity, Lifecycle.State.DESTROYED)

        // this guarantee that we will reinstall monitor between notifications about onDestroy
        // and onCreate
        // noinspection SynchronizationOnLocalVariableOrMethodParameter
        var result: T
        try {
            synchronized(monitor) {
                do {
                    // The documentation says "Block until an Activity is created
                    // that matches this monitor." This statement is true, but there are some other
                    // true statements like: "Block until an Activity is destroyed" or
                    // "Block until an Activity is resumed"...
                    // this call will release synchronization monitor's monitor
                    result =
                        monitor.waitForActivityWithTimeout(TIMEOUT_MS) as T
                            if (result == null) {
                        throw RuntimeException("Timeout. Activity was not recreated.")
                    }
                } while (result === activity)
            }
        } finally {
            instrumentation.removeMonitor(monitor)
        }

        // Finally wait for the recreated Activity to be resumed
        waitUntilState(result, Lifecycle.State.RESUMED)

        return result
    }
}
